src/main/java/com/amazon/ion/bytecode/BinaryPrimitiveReader.kt

```diff
@@ -0,0 +1,127 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.bytecode
+
+import com.amazon.ion.IonException
+import java.nio.ByteBuffer
+
+/**
+ * Helper class containing methods for reading FixedInts, and (TODO) FlexInts, FixedUInts, and FlexUInts.
+ *
+ * Some methods in this class use a technique to reduce the number of branches and minimize the number of calls to
+ * get data from the ByteBuffer. We know that any FlexInt or FlexUint must have at least 4 bytes preceding it (because
+ * the IVM is 4 bytes), so rather than reading bytes one at a time, we'll read one byte to figure out how many bytes to
+ * read, and then we'll read the entire FlexUInt (plus zero or more preceding bytes) in one call to [ByteBuffer.getInt]
+ * or [ByteBuffer.getLong]. This puts all the bytes we want into the _most_ significant bits of the `int` or `long`.
+ * Then we can remove the extra bytes and the continuation bits by using a single right-shift operation (signed for
+ * FlexInt or unsigned for FlexUInt). This technique significantly reduces the number of operations required to read a
+ * Flex(U)Int as compared to reading bytes one at a time.
+ *
+ * A similar technique is also used for reading FixedInts and FixedUInts.
+ *
+ * Examples:
+ * ```
+ * aaaa_aaaa bbbb_bbbb cccc_cccc dddd_dddd eeee_eee1 ffff_ffff gggg_gggg hhhh_hhhh iiii_iiii
+ * ```
+ * - read B-E... `eeee_eee1 dddd_dddd cccc_cccc bbbb_bbbb`
+ * - shift right by (8 * 3 + 1) = 25 = 4 + 7 * 3
+ * - unsigned shift right for FlexUint; signed shift right for FlexInt.
```

Comment:
I'm not sure these examples lead to an increased understanding of the code, which is fairly straightforward. I am having trouble grasping the concept that the examples are trying to convey with lines like `shift right by (8 * 2 + 2) = 18 = 4 + 7 * 2`, and yet I feel like I understand the code. Maybe the examples need to be rephrased or simply removed.